scipy/spatial/transform/_rotation_xp.py

```diff
@@ -0,0 +1,1138 @@
+    # find the composite angle from the weighted sum of the
+    # axial components in that plane, minimizing the 2D alignment
+    # problem.
+    sin_term = xp.linalg.vecdot(xp.linalg.cross(c_sec, a_sec), a_pri)
+    cos_term = xp.linalg.vecdot(c_sec, a_sec) - (
+        xp.linalg.vecdot(c_sec, a_pri) * xp.linalg.vecdot(a_sec, a_pri)
+    )
+    phi = xp.atan2(xp.sum(weights_sec * sin_term), xp.sum(weights_sec * cos_term))
+    q_sec = from_rotvec(phi * a_pri[0, ...])
+
+    # Compose these to get the optimal rotation
+    q_opt = xp.where(xp.asarray(N == 1), q_pri, compose_quat(q_sec, q_pri))
+
+    # Calculated the root sum squared distance. We force the error to
+    # be zero for the infinite weight vectors since they will align
+    # exactly.
+    weights_inf_zero = xp.asarray(weights, copy=True)
+
+    multiple_vectors = xp.asarray(N > 1, device=_device)
+    mask = xp.logical_or(multiple_vectors, weights[0] == xp.inf)
+    mask = xp.logical_and(mask, weight_is_inf)
+    # Skip non-infinite weight single vectors pairs, we used the
+    # infinite weight code path but don't want to zero that weight
+    weights_inf_zero = xpx.at(weights_inf_zero)[mask].set(0)
+    a_est = apply(q_opt, b)
+    rssd = xp.sqrt(xp.sum(weights_inf_zero @ (a - a_est) ** 2))
+
+    mask = xp.any(xp.isnan(weights), axis=-1)
+    q_opt = xp.where(mask, xp.nan, q_opt)
+    return q_opt, rssd, xp.asarray(xp.nan, device=_device)
+
+
+def pow(quat: Array, n: float) -> Array:
+    xp = array_namespace(quat)
+    _device = device(quat)
+    # general scaling of rotation angle
+    result = from_rotvec(n * as_rotvec(quat))
+    # Special cases 0 -> identity, -1 -> inv, 1 -> copy
+    identity = xp.zeros((*quat.shape[:-1], 4), dtype=result.dtype, device=_device)
+    identity = xpx.at(identity)[..., 3].set(1)
+    mask = xp.asarray(n == 0, device=_device)
+    result = xp.where(mask, identity, result)
+    mask = xp.asarray(n == -1, device=_device)
+    result = xp.where(mask, inv(quat), result)
+    mask = xp.asarray(n == 1, device=_device)
+    result = xp.where(mask, quat, result)
+    return result
+
+
+def _normalize_quaternion(quat: Array) -> Array:
+    xp = array_namespace(quat)
+    quat_norm = xp_vector_norm(quat, axis=-1, keepdims=True, xp=xp)
+    zero_norm = quat_norm == 0
+    if not is_lazy_array(quat_norm):
+        if xp.any(zero_norm):
+            raise ValueError("Found zero norm quaternions in `quat`.")
+    else:
+        quat = xp.where(zero_norm, xp.nan, quat)
+    return quat / quat_norm
+
+
+def _quat_canonical(quat: Array) -> Array:
+    xp = array_namespace(quat)
+    mask = quat[..., 3] < 0
+    zero_w = quat[..., 3] == 0
+    mask = xp.logical_or(mask, zero_w & (quat[..., 0] < 0))
+    zero_wx = xp.logical_or(zero_w, quat[..., 0] == 0)
+    mask = xp.logical_or(mask, zero_wx & (quat[..., 1] < 0))
+    zero_wxy = xp.logical_or(zero_wx, quat[..., 1] == 0)
+    mask = xp.logical_or(mask, zero_wxy & (quat[..., 2] < 0))
```

Comment:
```suggestion
    mask = quat[..., 3] < 0
    zero_w = quat[..., 3] == 0
    mask = xp.logical_or(mask, zero_w & (quat[..., 0] < 0))
    zero_wx = xp.logical_and(zero_w, quat[..., 0] == 0)
    mask = xp.logical_or(mask, zero_wx & (quat[..., 1] < 0))
    zero_wxy = xp.logical_and(zero_wx, quat[..., 1] == 0)
    mask = xp.logical_or(mask, zero_wxy & (quat[..., 2] < 0))
```
I believe these should be ands, we should update the tests if this did not get caught.

Reply:
Fixed and updated the tests